### ISSUE TITLE
Check if moize will retrieve a value from cache or will call a function

### ISSUE THREAD

**Issue by Vlad-Zhukov** [planttheidea/moize#39]
Hello! Thank you for the library, it's great!

I currently use it for caching promise results from data fetching. Each call to memoized function returns a Promise and starts a promise chain. While this is the correct and expected behaviour, practically it's not needed; usually you only want it to happen when the actual request happens.

Is there a way to test if `moize` will call a promise (e.g. new args, TTL expired or something else)? If there isn't what do you think about implementing one?

**Comment by planttheidea** [comment]
I'm not sure if I'm following this correctly, so forgive me if this answer doesn't tackle what you are looking for.

Currently `moize` will create a new `Promise` that is resolved with the value that is cached ... this ensures that you can treat the function return in the same way you would if you were calling it directly. If a function has a signature under normal circumstances that is a Promise, but a non-promise if cached, it creates a discrepancy that you need to bookkeep yourself. There is no delay in the cached response, its just a `Promise.resolve(value)` thing, so it is quite fast, but the interface remains consistent. 

The question at the end is a bit confusing ... there is no internal mechanism to determine if `moize` will handle promises or not, it is declared explicitly via the `isPromise` property on options, or on using `moize.promise`. If you just use the standard `moize` it will cache the `Promise` returned as a regular value.

Not sure if I'm answering your question or helping out ... perhaps some more clarity to the request would help me in my response.

**Comment by Vlad-Zhukov** [comment]
My question is not about promises specifically but they provide a very good example of side effects.

A simple example:
```js
const fn = (...args) => fetch('https://google.com/');

const memoized = moize(fn, {
  isPromise: true,
  maxAge: 60000,
});

// then somewhere later

// ... do something here before calling the promise

memoized(...args)
  .then(response => {
    // handle the response here
  })
  .catch(error => {
    // handle the error here
  });
```

Each time I call the `memoized` it will return a promise and run `then`, and as I said it's fine, but practically not useful. I want to run `memoized` only when it is guaranteed going to fetch new data, like so:

```js
if (/* if memoized will NOT retrieve from cache */) {
// ... do something here before calling the promise

memoized(...args)
  .then(response => {
    // handle the response here
  })
  .catch(error => {
    // handle the error here
  });
}
```

I guess that's what `cache.has` for, but it also requires `createGetCacheKey`. If would be nice to a method like this: `if(memoized.test(...args)) {...}`.

**Comment by planttheidea** [comment]
Ah i see ... your looking for some generic way to test if the key exists in cache or not, yes? Which currently is problematic with the multiple-parameter setup?

**Comment by Vlad-Zhukov** [comment]
Yeah, right. I'm sorry I overcomplicated such a simple question. I just thought it would be important to explain it from a real-world scenario perspective.

**Comment by planttheidea** [comment]
Yup I got ya, no worries. On it! Hopefully should have something in the next day or so.

**Comment by planttheidea** [comment]
I have published version `2.4.0`, which includes the new [hasCacheFor](https://github.com/planttheidea/moize#hascacheforargs) method. To use, pass the arguments to it as you would the function:

```javascript
const fn = (one, two, three) => {
  return one + two + three;
};

fn(1, 2, 3);

console.log(fn.hasCacheFor(1, 2, 3)); // true
console.log(fn.hasCacheFor(3, 2, 1)); // false
```

Let me know if this gives what you were looking for, and if so then I'll close the issue.

**Comment by Vlad-Zhukov** [comment]
This is exactly what I need, thanks!
